File: tests/data_structures/hash_table/hash_table.py
#!/usr/bin/env python3
"""
Hash table benchmark implementation in Python.
Tests hash table operations (insert, lookup, delete).
"""

import sys
import time
import random
import string


def random_string(length=10):
    """Generate a random string of given length."""
    return ''.join(random.choices(string.ascii_letters + string.digits, k=length))


def main():
    """Main execution function."""
    num_operations = 100000
    
    # Generate test data
    keys = [random_string() for _ in range(num_operations)]
    values = [random.randint(1, 1000000) for _ in range(num_operations)]
    
    print(f"Testing hash table with {num_operations} operations...")
    start_time = time.perf_counter()
    
    # Create hash table (dictionary in Python)
    hash_table = {}
    
    # Insert operations
    insert_start = time.perf_counter()
    for i in range(num_operations):
        hash_table[keys[i]] = values[i]
    insert_time = time.perf_counter() - insert_start
    
    # Lookup operations
    lookup_start = time.perf_counter()
    found_count = 0
    for key in keys:
        if key in hash_table:
            found_count += 1
    lookup_time = time.perf_counter() - lookup_start
    
    # Delete operations
    delete_start = time.perf_counter()
    deleted_count = 0
    for i in range(0, num_operations, 2):  # Delete every other key
        if keys[i] in hash_table:
            del hash_table[keys[i]]
            deleted_count += 1
    delete_time = time.perf_counter() - delete_start
    
    end_time = time.perf_counter()
    total_time = end_time - start_time
    
    print(f"Result:")
    print(f"  Inserted: {num_operations} items")
    print(f"  Found: {found_count}/{num_operations} items")
    print(f"  Deleted: {deleted_count} items")
    print(f"  Remaining: {len(hash_table)} items")
    print(f"Timing:")
    print(f"  Insert time: {insert_time:.6f} seconds")
    print(f"  Lookup time: {lookup_time:.6f} seconds") 
    print(f"  Delete time: {delete_time:.6f} seconds")
    print(f"  Total time: {total_time:.6f} seconds")


if __name__ == "__main__":
    main()

File: tests/data_structures/hash_table/hash_table.rs
use std::collections::HashMap;
use std::time::Instant;
use rand::{distributions::Alphanumeric, thread_rng, Rng};

fn random_string(length: usize) -> String {
    thread_rng()
        .sample_iter(&Alphanumeric)
        .take(length)
        .map(char::from)
        .collect()
}

fn main() {
    let num_operations = 100_000;
    
    // Generate test data
    let keys: Vec<String> = (0..num_operations)
        .map(|_| random_string(10))
        .collect();
    
    let values: Vec<i32> = (0..num_operations)
        .map(|_| thread_rng().gen_range(1..=1_000_000))
        .collect();
    
    println!("Testing hash table with {} operations...", num_operations);
    let total_start = Instant::now();
    
    // Create hash table
    let mut hash_table = HashMap::new();
    
    // Insert operations
    let insert_start = Instant::now();
    for i in 0..num_operations {
        hash_table.insert(keys[i].clone(), values[i]);
    }
    let insert_time = insert_start.elapsed();
    
    // Lookup operations
    let lookup_start = Instant::now();
    let found_count = keys.iter()
        .filter(|key| hash_table.contains_key(*key))
        .count();
    let lookup_time = lookup_start.elapsed();
    
    // Delete operations (every other key)
    let delete_start = Instant::now();
    let mut deleted_count = 0;
    for i in (0..num_operations).step_by(2) {
        if hash_table.remove(&keys[i]).is_some() {
            deleted_count += 1;
        }
    }
    let delete_time = delete_start.elapsed();
    
    let total_time = total_start.elapsed();
    
    println!("Result:");
    println!("  Inserted: {} items", num_operations);
    println!("  Found: {}/{} items", found_count, num_operations);
    println!("  Deleted: {} items", deleted_count);
    println!("  Remaining: {} items", hash_table.len());
    println!("Timing:");
    println!("  Insert time: {:.6} seconds", insert_time.as_secs_f64());
    println!("  Lookup time: {:.6} seconds", lookup_time.as_secs_f64());
    println!("  Delete time: {:.6} seconds", delete_time.as_secs_f64());
    println!("  Total time: {:.6} seconds", total_time.as_secs_f64());
}
